examples/10_recipe.py

```diff
@@ -0,0 +1,451 @@
+# need to parse and transform into proper datetime objects from which we
+# can then extract information.
+#
+# Without Skrub, we might do that on the whole dataframe using something like:
+
+# %%
+
+# Don't do this:
+# df["date_first_hired"] = pd.to_datetime(df["date_first_hired"])
+
+# %%
+# However, we would then need to remember to apply the same transformation when
+# we receive new data for which we are asked to make a prediction. And we would
+# also need to store the detected datetime format to be sure we apply the same
+# transformation. Skrub helps us integrate that data-wrangling operation into
+# our machine-learning model.
+#
+# To do so, we now add the first step to our recipe.
+
+# %%
+from skrub import ToDatetime
+
+recipe = recipe.add(ToDatetime(), cols="date_first_hired")
+recipe
+
+# %%
+# To add a transformation, we call ``recipe.add()`` with the transformer to
+# apply and optionally ``cols``, the set of columns it should modify.
+#
+# The columns can be a single column name (as shown above), or a list of column names.
+# Skrub also provides more powerful column selectors as we will see below.
+#
+# .. admonition:: ``recipe.add()`` returns a new object
+#
+#     Note that we wrote ``recipe = recipe.add(...)``. This is necessary because
+#     ``add()`` does not modify the recipe in-place. Instead, it returns a new
+#     |Recipe| object which contains the added step. To keep working with this new,
+#     augmented recipe, we bind the variable ``recipe`` to the new object.
+#
+# We can check what the data looks like with the added transformation. In the
+# report's **"Show: All columns"** dropdown, we can filter which columns we want
+# to see in the report. The default is to show all of them, but if we select
+# **"Modified by last step"**, we see only those that were created or modified by
+# the transformation we just added. In our case, this is the
+# ``"date_first_hired"`` column. We can see that it has a new dtype: it is now
+# a true datetime column.
+
+# %%
+recipe.get_report()
+
+# %%
+# More flexible column selection
+# ------------------------------
+#
+# Now that we have converted ``"date_first_hired"`` to datetimes, we can extract
+# temporal features such as the year and month with Skrub's |DatetimeEncoder|.
+#
+# We could again specify ``cols="date_first_hired"`` but we will take this
+# opportunity to introduce more flexible ways of selecting columns.
+#
+# The ``skrub.selectors`` module provides objects that we can use instead of
+# an explicit list of column names, and which allow us to select columns based
+# on their type, patterns in their name, etc. If you have used Polars or Ibis
+# selectors before, they work in the same way.
+#
+# For example, ``skrub.selectors.any_date()`` selects all columns that have a
+# datetime or date dtype.
+
+# %%
+from skrub import DatetimeEncoder
+from skrub import selectors as s
+
+recipe = recipe.add(DatetimeEncoder(), cols=s.any_date())
```

Comment:
we can (I guess should) very easily add a ToDatetime inside the DatetimeEncoder

they are 2 transformers because in the TableVectorizer they _have_ to be separate because the user provides the datetime encoder but does not control the column assignments. so the datetime columns must have been parsed to decide column assignments before they get assigned to the datetime encoder and reach it.

before, the main use case for datetime encoder was in the tablevectorizer. but now that it will become more practical to use without the tablevectorizer thanks to the recipe, adding datetime parsing to do it all in one go makes sense. (and the tablevectorizer just won't use this feature)

but there are also several other cleaning steps besides datetime parsing that the tablevectorizer does, so we might want either a transformer or an option for the recipe to apply all the cleaning / preprocessing (ie everything in the tablevectorizer except the user-provided final transformers)

Reply:
> we can (I guess should) very easily add a ToDatetime inside the DatetimeEncoder

That would be great IMO

> they are 2 transformers because in the TableVectorizer they have to be separate because the user provides the datetime encoder but does not control the column assignments. so the datetime columns must have been parsed to decide column assignments before they get assigned to the datetime encoder and reach it.

Yes, I remember the choices that led to this design, and I agree with them.

> before, the main use case for datetime encoder was in the tablevectorizer. but now that it will become more practical to use without the tablevectorizer thanks to the recipe, adding datetime parsing to do it all in one go makes sense. (and the tablevectorizer just won't use this feature)

Ok, if that doesn't introduce too much complexity on the TV part, I'm all for it.

> but there are also several other cleaning steps besides datetime parsing that the tablevectorizer does, so we might want either a transformer or an option for the recipe to apply all the cleaning / preprocessing (ie everything in the tablevectorizer except the user-provided final transformers)

That's interesting; I need to refresh my memory regarding this part.

Side question: Would using the TV with the Recipe make sense in general? I'm thinking about CVing the transformers and their hyper-parameters more easily.

Reply:
yes using the TableVectorizer in the Recipe completely makes sense, and it will help tune the choice of the encoders and their hyperparameters (the `choose_*` can be arbitrarily nested). I didn't do it in this example because on this dataset the TableVectorizer does everything fine so there would be only one step and it would be harder to showcase some features of the recipe

Reply:
Ok! What about mentioning the TV at the end? Or would that make the message less obvious?